sklearn_numba_dpex/kmeans/drivers.py

```diff
@@ -0,0 +1,290 @@
+import math
+
+import numpy as np
+import dpctl
+import pyopencl
+
+from sklearn_numba_dpex.utils._device import _DeviceParams
+
+from sklearn_numba_dpex.kmeans.kernels import (
+    make_initialize_to_zeros_1dim_float32_kernel,
+    make_initialize_to_zeros_2dim_float32_kernel,
+    make_initialize_to_zeros_3dim_float32_kernel,
+    make_copyto_kernel,
+    make_broadcast_division_kernel,
+    make_center_shift_kernel,
+    make_half_l2_norm_dim0_kernel,
+    make_sum_reduction_1dim_kernel,
+    make_sum_reduction_2dim_kernel,
+    make_sum_reduction_3dim_kernel,
+    make_fused_fixed_window_kernel,
+    make_assignment_fixed_window_kernel,
+)
+
+
+def _check_power_of_2(e):
+    if e != 2 ** (math.log2(e)):
+        raise RuntimeError(f"Expected a power of 2, got {e}")
+    return e
+
+
+class KMeansDriver:
+    def __init__(
+        self,
+        global_mem_cache_size=None,
+        preferred_work_group_size_multiple=None,
+        work_group_size_multiplier=None,
+        centroids_window_width_multiplier=None,
+        centroids_window_height_ratio_multiplier=None,
+        centroids_private_copies_max_cache_occupancy=None,
+        device=None,
+    ):
+        dpctl_device = dpctl.SyclDevice(device)
+        device_params = _DeviceParams(dpctl_device)
+
+        self.global_mem_cache_size = (
+            global_mem_cache_size or device_params.global_mem_cache_size
+        )
+
+        self.preferred_work_group_size_multiple = _check_power_of_2(
+            preferred_work_group_size_multiple
+            or device_params.preferred_work_group_size_multiple
+        )
+
+        self.work_group_size_multiplier = _check_power_of_2(
+            work_group_size_multiplier
+            or (
+                device_params.max_work_group_size // self.preferred_work_group_size_multiple
+            )
+        )
+
+        self.centroids_window_width_multiplier = _check_power_of_2(
+            centroids_window_width_multiplier or 1
+        )
+
+        self.centroids_window_height_ratio_multiplier = (
+            centroids_window_height_ratio_multiplier or 2
+        )
+
+        self.centroids_private_copies_max_cache_occupancy = (
+            centroids_private_copies_max_cache_occupancy or 0.7
+        )
+
+        self.device = device
+
+    def __call__(
+        self,
+        X,
+        sample_weight,
+        centers_init,
+        max_iter=300,
+        verbose=False,
+        x_squared_norms=None,
+        tol=1e-4,
+        n_threads=1,
+    ):
+        """Inspired by the "Fused Fixed" strategy exposed in:
+        Kruliš, M., & Kratochvíl, M. (2020, August). Detailed analysis and
+        optimization of CUDA K-means algorithm. In 49th International
+        Conference on Parallel Processing-ICPP (pp. 1-11).
```

Comment:
Always start a multiline docstring by a short oneliner summary (as per PEP-257).

```suggestion
        """GPU optimized implementation of Lloyd's k-means.

        Inspired by the "Fused Fixed" strategy exposed in:
        Kruliš, M., & Kratochvíl, M. (2020, August). Detailed analysis and
        optimization of CUDA K-means algorithm. In 49th International
        Conference on Parallel Processing-ICPP (pp. 1-11).
```

Reply:
(applied)